src/detection/terminalshell/terminalshell_linux.c

```diff
@@ -161,39 +166,47 @@ static pid_t getTerminalInfo(FFTerminalShellResult* result, pid_t pid)
 
     pid_t ppid = 0;
 
-    if(getProcessNameAndPpid(pid, name, &ppid))
-        return 0;
-
-    //Known shells
-    if (
-        strcasecmp(name, "ash")       == 0 ||
-        strcasecmp(name, "bash")      == 0 ||
-        strcasecmp(name, "zsh")       == 0 ||
-        strcasecmp(name, "ksh")       == 0 ||
-        strcasecmp(name, "mksh")      == 0 ||
-        strcasecmp(name, "oksh")      == 0 ||
-        strcasecmp(name, "csh")       == 0 ||
-        strcasecmp(name, "tcsh")      == 0 ||
-        strcasecmp(name, "fish")      == 0 ||
-        strcasecmp(name, "dash")      == 0 ||
-        strcasecmp(name, "pwsh")      == 0 ||
-        strcasecmp(name, "nu")        == 0 ||
-        strcasecmp(name, "git-shell") == 0 ||
-        strcasecmp(name, "elvish")    == 0 ||
-        strcasecmp(name, "oil.ovm")   == 0 ||
-        (strcasecmp(name, "python") == 0 && getenv("XONSH_VERSION"))
-    )
-        return getTerminalInfo(result, ppid);
+    while (getProcessNameAndPpid(pid, name, &ppid) == NULL)
+    {
+        //Known shells
+        if (
+            strcasecmp(name, "ash")       == 0 ||
+            strcasecmp(name, "bash")      == 0 ||
+            strcasecmp(name, "zsh")       == 0 ||
+            strcasecmp(name, "ksh")       == 0 ||
+            strcasecmp(name, "mksh")      == 0 ||
+            strcasecmp(name, "oksh")      == 0 ||
+            strcasecmp(name, "csh")       == 0 ||
+            strcasecmp(name, "tcsh")      == 0 ||
+            strcasecmp(name, "fish")      == 0 ||
+            strcasecmp(name, "dash")      == 0 ||
+            strcasecmp(name, "pwsh")      == 0 ||
+            strcasecmp(name, "nu")        == 0 ||
+            strcasecmp(name, "git-shell") == 0 ||
+            strcasecmp(name, "elvish")    == 0 ||
+            strcasecmp(name, "oil.ovm")   == 0 ||
+            (strcasecmp(name, "python") == 0 && getenv("XONSH_VERSION"))
+        )
+        {
+            pid = ppid;
+            continue;
+        }
 
-    #ifdef __APPLE__
-    // https://github.com/fastfetch-cli/fastfetch/discussions/501
-    if (ffStrEndsWith(name, " (figterm)") || ffStrEndsWith(name, " (cwterm)"))
-        getProcessNameAndPpid(ppid, name, &ppid);
-    #endif
+        #ifdef __APPLE__
+        // https://github.com/fastfetch-cli/fastfetch/discussions/501
+        if (ffStrEndsWith(name, " (figterm)") || ffStrEndsWith(name, " (cwterm)"))
+        {
+            if (getProcessNameAndPpid(ppid, name, &ppid) != NULL)
+                return 0;
```

Comment:
Do you really understand what the code does?

Reply:
> `ffReadFileData` would failed by no memory or other IO errors

In macOS we don't use `ffReadFileData`. But okey, this check makes at least no harm.